File: dp/python/last_stone_weight_ii.py
from typing import List


class Solution:
    def lastStoneWeightII(self, stones: List[int]) -> int:
        total = sum(stones)
        target = total // 2
        n = len(stones)
        # 初始化 dp[i][j]：可以放 0-i 物品，背包容量为 j 的情况下背包中的最大价值
        dp = [[0] * (target + 1) for _ in range(n)]
        # dp[i][0] 默认初始化为 0
        # dp[0][j] 取决于 stones[0]
        for j in range(stones[0], target + 1):
            dp[0][j] = stones[0]

        for i in range(1, n):
            for j in range(1, target + 1):  # 注意是等于
                if j >= stones[i]:
                    # 不放: dp[i - 1][j]  放: dp[i - 1][j - stones[i]] + stones[i]
                    dp[i][j] = max(dp[i - 1][j], dp[i - 1][j - stones[i]] + stones[i])
                else:
                    dp[i][j] = dp[i - 1][j]

        return (total - dp[n - 1][target]) - dp[n - 1][target]

File: dp/ts/last_stone_weight_ii.ts
// 1049. 最后一块石头的重量 II
// 二维数组方式：dp[i][j] 表示在石头下标 [0, i] 之间任取，凑出的总重量不超过 j 的最大值
function lastStoneWeightII(stones: number[]): number {
    const sum: number = stones.reduce((a: number, b: number): number => a + b);
    const target: number = Math.floor(sum / 2);
    const n: number = stones.length;
    // dp[i][j]：容量（背包能装的重量上限）为 j，只考虑前 i+1 块石头（下标 [0, i]）能装下的最大重量
    const dp: number[][] = Array.from({ length: n }, () => new Array(target + 1).fill(0));

    // 初始化第 0 块石头：能放得下就放
    for (let j: number = stones[0]; j <= target; j++) {
        dp[0][j] = stones[0];
    }

    // 背包递推
    for (let i: number = 1; i < n; i++) {
        for (let j: number = 0; j <= target; j++) {
            if (stones[i] > j) {
                // 放不下第 i 块，只能不放
                dp[i][j] = dp[i - 1][j];
            } else {
                // 不放 vs 放：取较大者
                dp[i][j] = Math.max(dp[i - 1][j], dp[i - 1][j - stones[i]] + stones[i]);
            }
        }
    }

    // 分成两组抵消，差值 = sum - 2 * 尽量接近 sum/2 的那一组
    return sum - dp[n - 1][target] - dp[n - 1][target];
}